Can build standalone but not as dependency

I'm trying to embed Tectonic in another program but am running into build issues that I can't quite pin down... 

I'm building on macOS with dependencies installed via Homebrew. 

This works just fine:
```
git clone https://github.com/tectonic-typesetting/tectonic
cd tectonic
git submodule update --init
PKG_CONFIG_PATH=$(brew --prefix icu4c)/lib/pkgconfig cargo build
```

But this fails:
```
cargo new tectonic-embed
cd tectonic-embed
cargo add tectonic
PKG_CONFIG_PATH=$(brew --prefix icu4c)/lib/pkgconfig cargo build
```

With trouble building the vendored HarfBuzz, it looks like. 
```
[snip]
error: failed to run custom build command for `tectonic_bridge_harfbuzz v0.2.9`

Caused by:
  process didn't exit successfully: `/Users/shane/Projects/tectonic-embed/target/debug/build/tectonic_bridge_harfbuzz-e78f2c9e3f07e8a3/build-script-build` (exit status: 101)
  --- stdout
  TARGET = Some("aarch64-apple-darwin")
  OPT_LEVEL = Some("0")
  HOST = Some("aarch64-apple-darwin")
  cargo:rerun-if-env-changed=CXX_aarch64-apple-darwin
  CXX_aarch64-apple-darwin = None
  cargo:rerun-if-env-changed=CXX_aarch64_apple_darwin
  CXX_aarch64_apple_darwin = None
  cargo:rerun-if-env-changed=HOST_CXX
  HOST_CXX = None
  cargo:rerun-if-env-changed=CXX
  CXX = None
  cargo:rerun-if-env-changed=CC_ENABLE_DEBUG_OUTPUT
  cargo:rerun-if-env-changed=CRATE_CC_NO_DEFAULTS
  CRATE_CC_NO_DEFAULTS = None
  DEBUG = Some("true")
  cargo:rerun-if-env-changed=CXXFLAGS_aarch64-apple-darwin
  CXXFLAGS_aarch64-apple-darwin = None
  cargo:rerun-if-env-changed=CXXFLAGS_aarch64_apple_darwin
  CXXFLAGS_aarch64_apple_darwin = None
  cargo:rerun-if-env-changed=HOST_CXXFLAGS
  HOST_CXXFLAGS = None
  cargo:rerun-if-env-changed=CXXFLAGS
  CXXFLAGS = None
  cargo:rerun-if-env-changed=AR_aarch64-apple-darwin
  AR_aarch64-apple-darwin = None
  cargo:rerun-if-env-changed=AR_aarch64_apple_darwin
  AR_aarch64_apple_darwin = None
  cargo:rerun-if-env-changed=HOST_AR
  HOST_AR = None
  cargo:rerun-if-env-changed=AR
  AR = None
  cargo:rerun-if-env-changed=ARFLAGS_aarch64-apple-darwin
  ARFLAGS_aarch64-apple-darwin = None
  cargo:rerun-if-env-changed=ARFLAGS_aarch64_apple_darwin
  ARFLAGS_aarch64_apple_darwin = None
  cargo:rerun-if-env-changed=HOST_ARFLAGS
  HOST_ARFLAGS = None
  cargo:rerun-if-env-changed=ARFLAGS
  ARFLAGS = None
  cargo:rustc-link-lib=static=harfbuzz
  cargo:rustc-link-search=native=/Users/shane/Projects/tectonic-embed/target/debug/build/tectonic_bridge_harfbuzz-d62862240405c468/out
  cargo:rerun-if-env-changed=CXXSTDLIB_aarch64-apple-darwin
  CXXSTDLIB_aarch64-apple-darwin = None
  cargo:rerun-if-env-changed=CXXSTDLIB_aarch64_apple_darwin
  CXXSTDLIB_aarch64_apple_darwin = None
  cargo:rerun-if-env-changed=HOST_CXXSTDLIB
  HOST_CXXSTDLIB = None
  cargo:rerun-if-env-changed=CXXSTDLIB
  CXXSTDLIB = None
  cargo:rustc-link-lib=c++
  cargo:include-path=/Users/shane/Projects/tectonic-embed/target/debug/build/tectonic_bridge_harfbuzz-d62862240405c468/out;/opt/homebrew/Cellar/graphite2/1.3.14/include

  --- stderr
  thread 'main' panicked at /Users/shane/.cargo/registry/src/index.crates.io-6f17d22bba15001f/tectonic_bridge_harfbuzz-0.2.9/build.rs:143:47:
  failed to copy header: Os { code: 2, kind: NotFound, message: "No such file or directory" }
  note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

I noted [this discussion](https://github.com/tectonic-typesetting/tectonic/issues/1172) where the problem was resolved by enabling `external-harfbuzz` but that just seems to have a different set of build issues where it looks like the XeTeX build is looking for `<harfbuzz/hb.h>` but the pkg-config include flag is already pointing to the `harfbuzz` directory so it's missing the header: 
```
[snip]
The following warnings were emitted during compilation:

warning: tectonic_xetex_layout@0.2.4: In file included from layout/xetex-XeTeXFontInst.cpp:42:
warning: tectonic_xetex_layout@0.2.4: layout/tectonic_xetex_layout.h:42:10: fatal error: 'harfbuzz/hb.h' file not found
warning: tectonic_xetex_layout@0.2.4: #include <harfbuzz/hb.h>
warning: tectonic_xetex_layout@0.2.4:          ^~~~~~~~~~~~~~~
warning: tectonic_xetex_layout@0.2.4: 1 error generated.

error: failed to run custom build command for `tectonic_xetex_layout v0.2.4`
```

I was about to start mucking around with CFLAGS but figured there might be some deeper issue I'm missing, at least with regard to getting the vendored version building properly when the Tectonic is a dependency.

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -97,6 +97,7 @@ watchexec-filterer-globset = "3.0"
 watchexec-signals = "2.0"
 watchexec-supervisor = "1.0"
 zip = { version = "^0.6", default-features = false, features = ["deflate"] }
+time = "0.3.36"
 
 [features]
 default = ["geturl-reqwest", "serialization"]

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -16,16 +16,13 @@ documentation = "https://docs.rs/tectonic"
 repository = "https://github.com/tectonic-typesetting/tectonic/"
 readme = "CARGO_README.md"
 keywords = ["tex", "latex", "typesetting", "font"]
-categories = [
-    "command-line-interface",
-    "parser-implementations",
-    "rendering",
-    "science",
-    "text-processing",
-]
+categories = ["command-line-interface", "parser-implementations", "rendering", "science", "text-processing"]
 license = "MIT"
 edition = "2018"
 exclude = ["/dist/", "/reference_sources/"]
+
+[lints.rust]
+unexpected_cfgs = { level = "warn", check-cfg = ['cfg(backtrace)'] }
 
 [badges]
 travis-ci = { repository = "tectonic-typesetting/tectonic" }
@@ -115,10 +112,7 @@ external-harfbuzz = ["tectonic_engine_xetex/external-harfbuzz"]
 geturl-curl = ["tectonic_bundles/geturl-curl", "tectonic_geturl/curl"]
 geturl-reqwest = ["tectonic_bundles/geturl-reqwest", "tectonic_geturl/reqwest"]
 
-native-tls-vendored = [
-    "tectonic_bundles/native-tls-vendored",
-    "tectonic_geturl/native-tls-vendored",
-]
+native-tls-vendored = ["tectonic_bundles/native-tls-vendored", "tectonic_geturl/native-tls-vendored"]
 
 # developer feature to compile with the necessary flags for profiling tectonic.
 profile = []
@@ -142,23 +136,9 @@ overlay-triplets-path = "dist/vcpkg-triplets"
 # guidance if they might need to set $VCPKGRS_TRIPLET.
 [package.metadata.vcpkg.target]
 x86_64-apple-darwin = { install = ["freetype", "harfbuzz[graphite2]", "icu"] }
-aarch64-apple-darwin = { triplet = "arm64-osx", install = [
-    "freetype",
-    "harfbuzz[graphite2]",
-    "icu",
-] }
-x86_64-unknown-linux-gnu = { install = [
-    "fontconfig",
-    "freetype",
-    "harfbuzz[graphite2]",
-    "icu",
-] }
-x86_64-pc-windows-msvc = { triplet = "x64-windows-static-release", install = [
-    "fontconfig",
-    "freetype",
-    "harfbuzz[graphite2]",
-    "icu",
-] }
+aarch64-apple-darwin = { triplet = "arm64-osx", install = ["freetype", "harfbuzz[graphite2]", "icu"] }
+x86_64-unknown-linux-gnu = { install = ["fontconfig", "freetype", "harfbuzz[graphite2]", "icu"] }
+x86_64-pc-windows-msvc = { triplet = "x64-windows-static-release", install = ["fontconfig", "freetype", "harfbuzz[graphite2]", "icu"] }
 
 [package.metadata.internal_dep_versions]
 tectonic_bridge_core = "thiscommit:2023-06-11:PvhF7YB"

--- a/crates/bundles/src/cache.rs
+++ b/crates/bundles/src/cache.rs
@@ -460,7 +460,7 @@ impl<CB: CacheBackend> CachingBundle<CB> {
         // If a filename contains newline characters, it will mess up our
         // line-based manifest format. Be paranoid and refuse to record such
         // filenames.
-        if !name.contains(|c| c == '\n' || c == '\r') {
+        if !name.contains(['\n', '\r']) {
             writeln!(man, "{name} {length} {digest_text}")?;
         }
 
@@ -515,9 +515,7 @@ impl<CB: CacheBackend> CachingBundle<CB> {
             // giving incorrect results if we pulled files out of the cache
             // before this invocation. Rewrite the digest file so that next time
             // we'll start afresh, then bail.
-            file_create_write(&self.digest_path, |f| {
-                writeln!(f, "{}", pull_data.digest.to_string())
-            })?;
+            file_create_write(&self.digest_path, |f| writeln!(f, "{}", pull_data.digest))?;
             bail!("backend digest changed; rerun tectonic to use updated information");
         }
 

--- a/crates/docmodel/src/document.rs
+++ b/crates/docmodel/src/document.rs
@@ -263,7 +263,7 @@ impl Document {
                 tried_src_path = true;
 
                 if let Some(s) = t.to_str() {
-                    name = s.to_owned();
+                    s.clone_into(&mut name);
                 }
             }
 
@@ -273,7 +273,7 @@ impl Document {
 
                     if let Some(Component::Normal(t)) = full_path.components().next_back() {
                         if let Some(s) = t.to_str() {
-                            name = s.to_owned();
+                            s.clone_into(&mut name);
                         }
                     }
                 }

--- a/crates/engine_spx2html/src/assets.rs
+++ b/crates/engine_spx2html/src/assets.rs
@@ -179,7 +179,7 @@ pub(crate) fn create_output_path(
 
     for piece in dest_path.split('/') {
         if let Some(out_path) = out_path.as_mut() {
-            match std::fs::create_dir(&out_path) {
+            match std::fs::create_dir(out_path.as_path()) {
                 Ok(_) => {}
                 Err(e) if e.kind() == std::io::ErrorKind::AlreadyExists => {}
                 Err(e) => {

--- a/crates/engine_spx2html/src/fontfile.rs
+++ b/crates/engine_spx2html/src/fontfile.rs
@@ -426,9 +426,10 @@ impl FontFileData {
 
         if let Some(out_path) = out_path.as_mut() {
             out_path.push(rel_path);
+            let display_path = out_path.clone();
             atry!(
-                std::fs::write(&out_path, &self.buffer);
-                ["cannot write output file `{}`", out_path.display()]
+                std::fs::write(out_path, &self.buffer);
+                ["cannot write output file `{}`", display_path.display()]
             );
         }
 
@@ -490,9 +491,10 @@ impl FontFileData {
 
                 out_path.pop();
                 out_path.push(&varname);
+                let display_path = out_path.clone();
                 atry!(
-                    std::fs::write(&out_path, &buffer);
-                    ["cannot write output file `{}`", out_path.display()]
+                    std::fs::write(out_path, &buffer);
+                    ["cannot write output file `{}`", display_path.display()]
                 );
             }
 

--- a/crates/engine_spx2html/src/fonts.rs
+++ b/crates/engine_spx2html/src/fonts.rs
@@ -197,13 +197,13 @@ impl FontEnsemble {
         // Update the info records for the relevant fonts to capture the
         // established relationship.
 
-        self.font_files[regular].family_name = name.clone();
+        self.font_files[regular].family_name.clone_from(&name);
         self.font_files[regular].family_relation = FamilyRelativeFontId::Regular;
-        self.font_files[bold].family_name = name.clone();
+        self.font_files[bold].family_name.clone_from(&name);
         self.font_files[bold].family_relation = FamilyRelativeFontId::Bold;
-        self.font_files[italic].family_name = name.clone();
+        self.font_files[italic].family_name.clone_from(&name);
         self.font_files[italic].family_relation = FamilyRelativeFontId::Italic;
-        self.font_files[bold_italic].family_name = name.clone();
+        self.font_files[bold_italic].family_name.clone_from(&name);
         self.font_files[bold_italic].family_relation = FamilyRelativeFontId::BoldItalic;
 
         self.font_families.insert(

--- a/crates/engine_spx2html/src/initialization.rs
+++ b/crates/engine_spx2html/src/initialization.rs
@@ -167,12 +167,12 @@ impl InitializationState {
     }
 
     fn handle_set_template(&mut self, texpath: &str, _common: &mut Common) -> Result<()> {
-        self.next_template_path = texpath.to_owned();
+        texpath.clone_into(&mut self.next_template_path);
         Ok(())
     }
 
     fn handle_set_output_path(&mut self, texpath: &str, _common: &mut Common) -> Result<()> {
-        self.next_output_path = texpath.to_owned();
+        texpath.clone_into(&mut self.next_output_path);
         Ok(())
     }
 

--- a/crates/engine_xetex/build.rs
+++ b/crates/engine_xetex/build.rs
@@ -25,6 +25,7 @@ fn main() {
     let profile_target_requires_frame_pointer: bool =
         target_cfg!(not(all(target_os = "linux", target_arch = "x86_64")));
 
+    #[allow(unexpected_cfgs)]
     const PROFILE_BUILD_ENABLED: bool = cfg!(feature = "profile");
 
     let profile_config = |cfg: &mut cc::Build| {

--- a/crates/io_base/src/digest.rs
+++ b/crates/io_base/src/digest.rs
@@ -4,7 +4,7 @@
 //! Helpers to tidy up the computation of digests in various places.
 
 use std::{
-    fs,
+    fmt, fs,
     path::{Path, PathBuf},
     str::FromStr,
 };
@@ -103,9 +103,9 @@ impl DigestData {
     }
 }
 
-impl ToString for DigestData {
-    fn to_string(&self) -> String {
-        bytes_to_hex(&self.0)
+impl fmt::Display for DigestData {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", bytes_to_hex(&self.0))
     }
 }
 

--- a/crates/xdv/src/lib.rs
+++ b/crates/xdv/src/lib.rs
@@ -1229,7 +1229,7 @@ impl<'a, T: XdvEvents> Cursor<'a, T> {
             return Err(InternalError::NeedMoreData);
         }
 
-        let rv = unsafe { mem::transmute(self.buf[0]) };
+        let rv = unsafe { mem::transmute::<u8, i8>(self.buf[0]) };
         self.buf = &self.buf[1..];
         self.offset += 1;
         Ok(rv)

--- a/crates/xetex_layout/build.rs
+++ b/crates/xetex_layout/build.rs
@@ -86,7 +86,7 @@ fn main() {
     let mut cppcfg = cc::Build::new();
 
     let cppflags = [
-        "-std=c++14",
+        "-std=c++17",
         "-Wall",
         "-Wdate-time",
         "-Wendif-labels",

--- a/src/driver.rs
+++ b/src/driver.rs
@@ -1132,7 +1132,7 @@ impl ProcessingSessionBuilder {
                     }
                 };
 
-                filesystem_root = parent.clone();
+                filesystem_root.clone_from(&parent);
                 let pio: Box<dyn IoProvider> = Box::new(FilesystemPrimaryInputIo::new(&p));
                 (pio, Some(p), parent)
             }

--- a/src/io/format_cache.rs
+++ b/src/io/format_cache.rs
@@ -55,7 +55,7 @@ impl FormatCache {
         let mut p = self.formats_base.clone();
         p.push(format!(
             "{}-{}-{}.fmt",
-            self.bundle_digest.to_string(),
+            self.bundle_digest,
             stem,
             crate::FORMAT_SERIAL
         ));

--- a/tests/formats.rs
+++ b/tests/formats.rs
@@ -176,12 +176,8 @@ fn test_format_generation(texname: &str, fmtname: &str, sha256: &str) {
             let observed = info.write_digest.unwrap();
 
             if observed != want_digest {
-                println!(
-                    "expected {} to have SHA256 = {}",
-                    fmtname,
-                    want_digest.to_string()
-                );
-                println!("instead, got {}", observed.to_string());
+                println!("expected {} to have SHA256 = {}", fmtname, want_digest);
+                println!("instead, got {}", observed);
                 panic!();
             }
         }

--- a/tests/util/mod.rs
+++ b/tests/util/mod.rs
@@ -216,7 +216,7 @@ impl<'a> ExpectedFile<'a> {
             .unwrap_or_else(|| panic!("couldn't Unicode-ify file name of {:?}", pbase))
             .replace(std::path::is_separator, "_");
 
-        let mut dec = GzDecoder::new(File::open(&pbase).unwrap());
+        let mut dec = GzDecoder::new(File::open(pbase.as_path()).unwrap());
         let mut contents = Vec::new();
         dec.read_to_end(&mut contents).unwrap();
 

